scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,25 +1089,72 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
-      }
+      val prepared  = mergeAvoidingHashes(merged)
+      prepared
+    }
+  }
+
+  /**
+    * Special handling for cases where one side of the hashjoin is merged
+    *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+    *  so we insert checkpoints and/or intermediate merge stages as appropriate
+    *
+    * @param pipes a list of pipes possibly colliding with each other
+    * @return an updated list of pipes, which can safely be merged
+    */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
+    require(!pipes.isEmpty)
+
+    // we make use of the fact that the pipe merge operation is not just associative but also commutative
+    val (colliding, uncolliding) = pipes.partition(p => pipes.find(o => (o != p) && RichPipe.isHashJoinedWithPipe(p, o)).isDefined)
```

Comment:
OK, I tried something. Now outputs ```new Merge(new Merge(a,c), b, d)``` in the example, still passes the Complex and the Twisted tests